### DISCUSSION TITLE
function "promptStringOnce" not defined

### DISCUSSION THREAD

**Discussion by papanito** [twpayne/chezmoi#2260] [Q&A] [answered]
I have a template file `dot_env.tmpl`

```
{{- $token_xraynop := promptStringOnce . "token_xraynop" "Token XRAY NOP" -}}
export XRAY_TOKEN_NOP={{ $token_xraynop | quote }}
```

However, when I run `chezmoi apply` I get the following error

```
chezmoi: template: dot_env.tmpl:1: function "promptStringOnce" not defined
```

Same happens if I use `promptSting`. I am currently running this on a `MINGW64_NT-10.0-17763`

**Comment by papanito** [comment]
It works fine, if it's defined in `.chezmoi.toml.tmpl`

**Comment by twpayne** [comment, marked as answer]
This is by design. Some functions are only available when [generating the initial config file](https://www.chezmoi.io/reference/templates/init-functions/). This is because you don't want to be prompted every time chezmoi executes a template (e.g. which happens when you run any of the `apply`, `diff`, or `status` commands, amongst others).

**Reply by papanito** [reply]
I though so, but thanks for clarifying

**Reply by grimm26** [reply]
So as I am starting out and I have run `chezmoi init` to begin, I develop a templated `.config/chezmoi/chezmoi.toml` with a `promptStringOnce` for email. How do I use this or at least stop it from saying `chezmoi: .config/chezmoi/chezmoi.toml: template: dot_config/chezmoi/chezmoi.toml.tmpl:1: function "promptStringOnce" not defined` every time I run a chezmoi command?

**Reply by halostatue** [reply]
Please reread the section on .chezmoi.toml. You aren't keeping in the right place and that won't work. This will likely become an error soon because trying to manage your chezmoi config this way does not work.

**Reply by grimm26** [reply]
@halostatue 
https://www.chezmoi.io/reference/configuration-file/ 

```
In most installations, the config file will be read from $HOME/.config/chezmoi/chezmoi.$FORMAT
(%USERPROFILE%/.config/chezmoi/chezmoi.$FORMAT), where $FORMAT is one of json, jsonc, toml, or yaml.
```

**Reply by grimm26** [reply]
aha, https://www.chezmoi.io/user-guide/setup/#create-a-config-file-on-a-new-machine-automatically $HOME/.config/chezmoi/chezmoi.$FORMAT is generated from .chezmoi.$FORMAT.tmpl

**Reply by halostatue** [reply]
Correct. This came up last week in another discussion and @twpayne (who I believe is on vacation) indicated that while we have protection against `chezmoi add ~/.config/chezmoi/chezmoi.toml` we don't have protection for `chezmoi apply` but should probably do so (and refuse to write the conflicting file).

This might also be something for `chezmoi doctor` to detect. I'm not in a position to write this myself right now.

**Reply by twpayne** [reply]
> This might also be something for `chezmoi doctor` to detect. I'm not in a position to write this myself right now.

There's already a "suspicious entries" check in `chezmoi doctor` for this.
